File: 04-Binary-Search-Tree/Code-Java/BinarySearch.java
package binarySearchTree;

import java.util.Arrays;

/**
 * 非递归的二分查找算法
 */
public class BinarySearch {
    // 在有序序列 lst 中, 查找 target
    // 如果找到 target, 返回相应的索引 index，否则返回 -1
    // 非递归方法
    public static int find(Comparable[] arr, Comparable target) {
        int l = 0;
        int r = arr.length;
        while (l <= r) {
            // 防止整形溢出，不使用 int mid = (l + r) / 2
            int mid = l + (r - l) / 2;
            if (arr[mid].compareTo(target) == 0)
                return mid;
            if (arr[mid].compareTo(target) > 0)
                r = mid - 1;
            else
                l = mid + 1;
        }
        return -1;
        // TODO:
    }

    // 递归方法
    public static int find2(Comparable[] arr, Comparable target, int l, int r) {
        if (l > r) return -1;
        int mid = l + (r - l) / 2;
        if (arr[mid].compareTo(target) == 0)
            return mid;
        else if (arr[mid].compareTo(target) > 0)
            return find2(arr, target, l, mid - 1);
        else return find2(arr, target, mid + 1, r);
    }
}

File: codePython/04-Binary-Search-Tree/binary_search.py
"""
二分查找算法
"""


def find1(lst, target):
    """
    在有序序列 lst 中, 查找 target
    如果找到 target, 返回相应的索引 index，否则返回 -1
    非递归方法
    """
    l, r = 0, len(lst) - 1
    while l <= r:
        # 防止整形溢出，不使用 mid = (l + r) // 2
        mid = l + (r - l) // 2
        if lst[mid] == target:
            return mid
        if lst[mid] > target:
            r = mid - 1
        else:
            l = mid + 1
    return -1


def find2(lst, target, l, r):
    """递归方法"""
    if l > r:
        return -1
    mid = l + (r - l) // 2
    if lst[mid] == target:
        return mid
    elif lst[mid] > target:
        return find2(lst, target, l, mid - 1)
    else:
        return find2(lst, target, mid + 1, r)

